packages/contracts/test/core/acl/integration-tests/parameter-scoping-oracle.ts

```diff
@@ -0,0 +1,96 @@
+import chai, {expect} from 'chai';
+import {ethers} from 'hardhat';
+import chaiUtils from '../../../test-utils';
+import {customError} from '../../../test-utils/custom-error-helper';
+import {SignerWithAddress} from '@nomiclabs/hardhat-ethers/signers';
+import {
+  TestParameterScopingACLOracle,
+  TestComponent,
+  DAO,
+} from '../../../../typechain';
+
+chai.use(chaiUtils);
+
+const DO_SOMETHING_ROLE = ethers.utils.id('DO_SOMETHING_ROLE');
+
+describe('TestParameterScopingOracle', function () {
+  let signers: SignerWithAddress[];
+  let parameterOracle: TestParameterScopingACLOracle;
+  let testComponent: TestComponent;
+  let dao: DAO;
+  let ownerAddress: string;
+
+  before(async () => {
+    signers = await ethers.getSigners();
+    ownerAddress = await signers[0].getAddress();
+
+    // create a DAO
+    const DAO = await ethers.getContractFactory('DAO');
+    dao = await DAO.deploy();
+    await dao.initialize('0x', ownerAddress, ethers.constants.AddressZero);
+
+    // create a component
+    const TestComponent = await ethers.getContractFactory('TestComponent');
+    testComponent = await TestComponent.deploy();
+    await testComponent.initialize(dao.address);
+
+    // create parameter oracle
+    const ParameterOracle = await ethers.getContractFactory(
+      'TestParameterScopingACLOracle'
+    );
+    parameterOracle = await ParameterOracle.deploy();
+
+    // Give signers[0] the DO_SOMETHING_ROLE on the TestComponent
+    dao.grantWithOracle(
+      testComponent.address,
+      ownerAddress,
+      DO_SOMETHING_ROLE,
+      parameterOracle.address
+    );
+  });
+
+  describe('oracle conditions:', async () => {
+    it('adds if the first parameter is larger than the second', async () => {
+      let param1 = 10;
+      let param2 = 1;
+
+      expect(
+        await testComponent.callStatic.addPermissioned(param1, param2)
+      ).to.be.equal(11);
+    });
+
+    it('reverts if the oracle is called by the wrong function', async () => {
+      let param1 = 10;
+      let param2 = 1;
+
+      await expect(
```

Comment:
Is this the intended way to test a revert? Can't we send an actual transaction?

Reply:
Yes, we use `callStatic` in many places where we are testing non-view functions and don't need the state changes to happen for subsequent tests. This is often also mandatory because the result from the function call is sometimes inaccessible because there is no getter or no event being emitted.

Reply:
I see, but testing the return value feels like only part of the thing. We may also test that the state didn't actually change after sending the transaction attempt. In cases where it is overkill we may do fine with `callStatic`.